mutation/async_utils.cc

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <seastar/core/coroutine.hh>
+
+#include "mutation/async_utils.hh"
+#include "mutation/mutation_partition_view.hh"
+#include "mutation/partition_version.hh"
+#include "partition_builder.hh"
+#include "mutation/canonical_mutation.hh"
+#include "converting_mutation_partition_applier.hh"
+#include "mutation/mutation_partition_serializer.hh"
+#include "idl/mutation.dist.impl.hh"
+
+future<> apply_gently(mutation_partition& target, const schema& s, mutation_partition_view p,
+        const schema& p_schema, mutation_application_stats& app_stats) {
+    mutation_partition p2(target, mutation_partition::copy_comparators_only{});
+    partition_builder b(p_schema, p2);
+    co_await p.accept_gently(p_schema, b);
+    if (s.version() != p_schema.version()) {
+        p2.upgrade(p_schema, s);
+    }
+    apply_resume res;
+    while (target.apply_monotonically(s, std::move(p2), no_cache_tracker, app_stats, is_preemptible::yes, res) == stop_iteration::no) {
+        co_await yield();
+    }
```

Comment:
Shouldn't this have a special case when the schemas match?

Reply:
You mean for correctness, or for performance?
The implementation here just follows `void mutation_partition::apply(const schema& s, mutation_partition_view p, const schema& p_schema, mutation_application_stats& app_stats)`.

In theory, it might be possible to combine `p.accept_gently` using the `partition_builder` with `apply_monotonically` to apply the `mutation_partition_view` directly into `target` if the schemas match, so we won't need this temporary `partition_builder`, but it's not trivial to say the least.  We should look at it as a possible optimization as a follow up.